### ISSUE TITLE
symlinks plugin move to new path , fisherman can't remove it

### ISSUE THREAD

**Issue by xrain0610** [fisherman/fisherman#182]
step 1: install the local plugin, 
```
fisher i ~/Desktop/webdev-docker
```
it like this:
![qq20160331-0 2x](https://cloud.githubusercontent.com/assets/11401602/14179738/da9f0c94-f792-11e5-8c0e-328775e7a373.png)

step 2: move the plugin to new path, and reinstall it ,
```
 fisher i ~/Documents/Fisherman/webdev-docker
```
it like this:
![qq20160331-1 2x](https://cloud.githubusercontent.com/assets/11401602/14179757/e7f6f096-f792-11e5-8fa1-f7807171bcba.png)

step3: fisher uninstall webdev-docker, but you can't remove all of them.

suggest:
1. when uninstall the plugin, if symlinks source not found, then only remove it from list

**Comment by bucaran** [comment]
@jethrokuan Are you able to repro this?

**Comment by jethrokuan** [comment]
@bucaran not exactly, but the plugin requires reinstallation after moving. I'm unable to get two plugins of the same name installed.

**Comment by detailyang** [comment]
I have repro this. The reason for this is we use `test -e $path` to judge whether symbolic link file exists without `test -L $path` and use `command ln -sFf ` to overwrite an existing symbolic. 

**Should we think `~/Documents/Fisherman/webdev-docker` and `~/Desktop/webdev-docker` is different local plugin?**

**Comment by detailyang** [comment]
I have repro this. The reason for this is we use `test -e $path` to judge whether symbolic link file exists without `test -L $path` and use `command ln -sFf ` to overwrite an existing symbolic. 

**Should we think `~/Documents/Fisherman/webdev-docker` and `~/Desktop/webdev-docker` is different local plugin?**

**Comment by bucaran** [comment]
Maybe someone can explain to me why, step 1: install the local plugin, and then step 2: move the plugin to new path, and reinstall it. 

Why do you need to move the plugin to a new path?

**Comment by detailyang** [comment]
+1 :+1: 
Why do we need to move the same plugin to a new path?

**Comment by bucaran** [comment]
I am still not sure I understand what's going, so let me describe how the link process _should_ work. If this is not how it presently works then this may be a bug.

```
mkdir my_plugin
cd my_plugin
$EDITOR my_plugin.fish
fisher i .
my_plugin
# success
```

To make  `my_plugin` available, we soft link `my_plugin` in the configuration.

That's all.

**Comment by detailyang** [comment]
yup, when soft link my_plugin in the configuration, we check file exists use -e, but when when symbolic link target_file has moved test -e $path will return false, so we also use -L to check whether is link
a little try for this pull request #192

**Comment by bucaran** [comment]
I still don't get it lol 

@jethrokuan Can you explain what's all this?

**Comment by jethrokuan** [comment]
I think what they're saying is after the soft link is made, and someone moves the folder containing the plugin, it breaks, and we should be checking if the soft link is still valid, and if it's not, remove it.

The question is not why someone would move the plugin folder around, but why fisherman breaks when it does.

**Comment by bucaran** [comment]
Okay, that makes sense. Would you mind double checking #192 too?

**Comment by detailyang** [comment]
agree with @jethrokuan。 `checking if the soft link is still valid` is a solution:)

**Comment by jethrokuan** [comment]
I don't think the PR solves the problem.

The `-e` check covers the case for broken soft links as well. It returns false if the soft link is broken.

**Comment by bucaran** [comment]
@jethrokuan Mmm, I think my brain is slow tonight, would you (or anyone) mind typing in the commands that I need to enter to repro the issue.

If I can understand the problem I am sure I could try to find as solution.

**Comment by detailyang** [comment]
yeah, The -e check covers the case for broken soft links as well, but the "soft link" iteself exists.
now master branch will overwrites the soft link without delete `~/Desktop/webdev-docker` on `$fisher_file`.
when  xrain0610 exec `fisher i ~/Documents/Fisherman/webdev-docker`, the fisher_file and fisher_cache as follow

>$fisher_file
<img width="677" alt="2016-04-10 00 02 30" src="https://cloud.githubusercontent.com/assets/3370345/14404740/2fe81f28-feb0-11e5-8faf-bac7ef2c6146.png">

>$fisher_cache
<img width="1278" alt="2016-04-10 00 01 57" src="https://cloud.githubusercontent.com/assets/3370345/14404743/44de12ca-feb0-11e5-836c-66fd33c32f4b.png">


anyway, i agree jethrokuan  `checking if the soft link is still valid`

**Comment by detailyang** [comment]
@bucaran  the command as follow i guess can repro this issue 
````bash
git clone https://github.com/xrain0610/webdev-docker ~/Desktop/webdev-docker
fisher i ~/Desktop/webdev-docker
mkdir -p ~/Documents/Fisherman/
mv ~/Desktop/webdev-docker/ ~/Documents/Fisherman/
fisher i ~/Documents/Fisherman/webdev-docker
````

**Comment by bucaran** [comment]
@detailyang Thanks, finally I get it.

Now, why would anyone try do that? I can imagine someone doing that by mistake, but is there a use case for this?

**Comment by detailyang** [comment]
@bucaran I'm not sure whether it's a bug. It's a strange behavior...But it's what @xrain0610 did.
should we solve this case?

**Comment by bucaran** [comment]
Yeah, let's solve this of course.

**Comment by detailyang** [comment]
let me try again:) i will send a pull request

**Comment by detailyang** [comment]
@bucaran @bucaran i update #192 , can double check this? thx :grin:

**Comment by jethrokuan** [comment]
Like I mentioned, it's not so much a use case as it is fisherman not being resilient to this kind of "mistakes".

PR looks good now, initially it was `test ! -e $path -a ! -L $path` where the `-L` would always be short-circuited.

**Comment by pickfire** [comment]
@jethrokuan, is there any reason to use `-L $path`?

I think that just `test -e $path; or delete_the_line` will do, this can also help to check if someone modified the fishfile.

**Comment by jethrokuan** [comment]
@pickfire that would work theoretically, but I think we want to be more careful with when things get deleted. if `test -e $path` fails and it's not a symlink, it signifies a bigger issue, and should not be kept hidden.

**Comment by detailyang** [comment]
+1 @jethrokuan  if test -e $path fails and it's not a symlink and the plugin is "installed",   it signifies a bigger issue.
should we tell user more detail about what we will to or tell user there is a broken link?

**Comment by bucaran** [comment]
The link feature is mostly aimed at developers prototyping plugins, but I guess users can use it to create aliases or plugins for local use only. Either way, the user is likely to be a developer. Should we really clean up user mistakes like this?

**Comment by jethrokuan** [comment]
I think users should clean up after their own mistakes here, but they should be made aware when something has gone awry.

**Comment by detailyang** [comment]
so we should tell user we will remove the broken link?

**Comment by bucaran** [comment]
Agree with @jethrokuan. 

> so we should tell user we will remove the broken link?

I'd say yes to this.

**Comment by detailyang** [comment]
how about this?
````bash
if test -L "$path"
    set -l realpath (command readlink "$path")
    if __fisher_file_remove "$realpath" "$fisher_file"
        printf "Remove broken plugin %s(%s) ok" "$name" "$realpath"
    else
        printf "Remove broken plugin %s(%s) fail" "$name" "$realpath"
    end
end
````

**Comment by bucaran** [comment]
LGTM

Let's use `debug` instead of `printf`. I'll merge it right away.

**Comment by detailyang** [comment]
@bucaran
 have used debug instead of printf on the PR #192 :smile:
